### ISSUE TITLE
Can't use fulltext search with traversing

### ISSUE THREAD

**Issue by StarpTech** [orientechnologies/orientdb#3877]
Is any support planned?

Example case:

```
select from (traverse out() from <rid> )  where <property> LUCENE "test*"
```

**Comment by lvca** [comment]
This is already supported

**Comment by StarpTech** [comment]
Can you give me an example?  Thanks.

**Comment by lvca** [comment]
The query above is supported, but the LUCENE filter happens in the outer query. What's your purpose?

**Comment by StarpTech** [comment]
I want to filter the traversed elements with the fulltext query. The query above dont use any index.

**Comment by lvca** [comment]
@StarpTech Sorry I go it now. LUCENE filters only with a FROM clause. In your case you could execute the LUCENE against the entire set and then filter by RID. Like:

    select from ( select from <your-class> where <property> LUCENE "test*" ) where <rid> IN in()

**Comment by StarpTech** [comment]
Thanks,  I will try it.

**Comment by StarpTech** [comment]
I got it but this will be very inefficient on large datasets.

```
select from Node where @rid IN ( select from ( traverse out() from (select from Node where Email = 'julia@local') ) ) and DisplayName LUCENE "dust*"
```

In my opinion this is a fundamental feature and should be improved.

**Comment by ncreea** [comment]
> I got it but this will be very inefficient on large datasets.
> In my opinion this is a fundamental feature and should be improved.

I have run into the same issue a while and did the same workaround @lvca mentioned, but soon I'll be in trouble, because the database is growing and in some situations this IN operation cost ~70% off all the processing time

**Comment by StarpTech** [comment]
@ncreea yes , that's terrible and is not acceptable for a graph database. These kind of bottlenecks make a database unusable for professional production. It would be nice if orientdb is more focussed to improve the existing features instead to blowup the db with all awesome stuff. The user @mattaylor said this in good words.

> Nice idea, but in my opinion it should be a third party plugin, not a part of the core.
I think that OrientDB has to be focused on its main purpose: become the most reliable, well documented, modular graph database, without wasting efforts in developing accessories before the main goal is achieved.
https://github.com/orientechnologies/orientdb/issues/2652#issuecomment-51158776

**Comment by lvca** [comment]
@StarpTech We did a POC a few months ago and this solution scaled quite good even with tens millions of entries. Did you execute a benchmark to see when this become a bottleneck?

**Comment by StarpTech** [comment]
@lvca @ncreea is talking about his real world scenario. It is obvious that this query isn't perfect for scalability. In general I didn't only talk about this example I am talking about the fact that there are much situations in orientdb which needs a lot improvement to be usable in production. Just to mention a few of them:

- Export/Import (no way to use this tools efficiently in production)
- A lot of bugs in the sql parser (one space more and the query is broken)
- ...

I like orientdb very much but the focus should be on the core features.

**Comment by lvca** [comment]
About this scaling issue by having one index for all the entries, we need proves about bottlenecks and in which cases. Can you measure index lookup with 1M or 1B items?
About the rest:
- Export/Import, what's the problem?
- In 2.1 the new sql parser fixed this noisy problem with spaces. 2.1-rc1 will be out tomorrow (but it's already available as 2.1-SNAPSHOT)
- We're working also on this
- Can you provide just one example?

**Comment by StarpTech** [comment]
**Export/import:**
Can you tell me one method to export my db  within an adequate time span and without locking my production environment?


**SqlParser**
`find references (select from Node where Email = 'julia@local')` works as expected.
`find references  (select from Node where Email = 'julia@local')` don't work (one space more after references keyword)

**Projection**
If I receive data with a fetchPlan with projection the nested records aren't use the projection before.
